File: leetcode/src/ArrayAndString/_30_Substring_with_Concatenation_of_All_Words.py
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        ans = []

        wordLen = len(words[0])
        counts = {}
        for word in words:
            counts[word] = counts.get(word, 0) + 1

        for i in range(len(s)):
            copy = {}
            isSame = True
            j = i + wordLen*len(words)
            if j > len(s):
                continue
            k = i
            while k+wordLen<=j:
                key = s[k: k+wordLen]
                if counts.has_key(key) is False or counts.get(key)<=copy.get(key):
                    isSame = False
                    break
                else:
                    copy[key] = copy.get(key, 0)+1
                k += wordLen

            if isSame is True:
                ans.append(i)
        return ans


Solution().findSubstring("barfoothefoobarman", ["foo","bar"])

File: leetcode/src/ArrayAndString/_30_Substring_with_Concatenation_of_All_Words.java
package ArrayAndString;

import java.util.*;

/**
 * Created by tangjialiang on 17/5/10.
 */
public class _30_Substring_with_Concatenation_of_All_Words {

    public static void main(String[] args) {
        String s = "wordgoodgoodgoodbestword" ;
        String[] words = {"word","good","best","good"} ;
        List<Integer> work = new _30_Substring_with_Concatenation_of_All_Words().work(s, words);
        for (Integer i : work) System.out.println(i);
    }

    private List<Integer> work(String s, String[] words) {
        List<Integer> substring = new Solution().findSubstring(s, words);
        return substring ;
    }

    public class Solution {
        ArrayList<Integer> ans = new ArrayList<Integer>() ;

        List<Integer> findSubstring(String s, String[] words) {
            StringBuilder sb = new StringBuilder(s) ;
            int wordLen = words[0].length() ;

            HashMap<String, Integer> counts = new HashMap<>() ;
            for (String word : words) {
                counts.put(word, (counts.containsKey(word))?(counts.get(word) + 1):(1)) ;
            }

            for(int i=0; i<s.length(); i++) {
                HashMap<String, Integer> copy = new HashMap<>() ;
                boolean is = true ;
                int j = i + wordLen*words.length ;
                if (j > s.length()) continue ;
                for (int k = i; k+wordLen <= j; k+=wordLen) {
                    // System.out.println("i: "+i+""+" j: "+j+" k: "+k) ;
                    String key = sb.substring(k, k+wordLen) ;
                    if (counts.get(key)==null || counts.get(key).equals(copy.get(key))) {
                        // not exits or not enough
                        is = false ;
                        break ;
                    } else {
                        // can
                        copy.put(key, (copy.get(key)!=null)?(copy.get(key)+1):(1)) ;
                    }
                }

                if (is) {
                    ans.add(i) ;
                }
            }

            return ans ;
        }
    }
}
